GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m

```diff
@@ -323,17 +350,128 @@ - (void)removeEvents:(NSSet<NSNumber *> *)eventIDs {
   });
 }
 
-#pragma mark - GDTCORStorageProtocol
+- (nullable NSNumber *)
+    batchWithEventSelector:(nonnull GDTCORStorageEventSelector *)eventSelector
+           batchExpiration:(nonnull GDTCORClock *)expiration
+                onComplete:(nonnull void (^)(NSNumber *_Nullable batchID,
+                                             NSSet<GDTCOREvent *> *_Nullable events))onComplete {
+  NSNumber *batchID = [self nextBatchID];
+  if (batchID == nil) {
+    return nil;
+  }
+  dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+  __block NSSet<NSString *> *eventPaths;
+  [self pathsForTarget:eventSelector.selectedTarget
+              eventIDs:eventSelector.selectedEventIDs
+              qosTiers:eventSelector.selectedQosTiers
+            mappingIDs:eventSelector.selectedMappingIDs
+            onComplete:^(NSSet<NSString *> *_Nonnull paths) {
+              eventPaths = paths;
+              dispatch_semaphore_signal(sema);
+            }];
+  if (dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_SEC)) != 0) {
+    return nil;
+  }
+  if (eventPaths.count == 0) {
+    return nil;
+  }
+  dispatch_async(_storageQueue, ^{
+    NSMutableSet<GDTCOREvent *> *events = [[NSMutableSet alloc] init];
+    for (NSString *eventPath in eventPaths) {
+      NSError *error;
+      GDTCOREvent *event =
+          (GDTCOREvent *)GDTCORDecodeArchive([GDTCOREvent class], eventPath, nil, &error);
+      if (event == nil || error) {
+        GDTCORLogDebug(@"Error deserializing event: %@", error);
+        [[NSFileManager defaultManager] removeItemAtPath:eventPath error:nil];
+        continue;
+      } else {
+        NSString *fileName = [eventPath lastPathComponent];
+        NSString *batchPath = [GDTCORFlatFileStorage batchPathForTarget:eventSelector.selectedTarget
+                                                                batchID:batchID];
+        [[NSFileManager defaultManager] createDirectoryAtPath:batchPath
+                                  withIntermediateDirectories:YES
+                                                   attributes:nil
+                                                        error:nil];
+        NSString *destinationPath = [batchPath stringByAppendingPathComponent:fileName];
```

Comment:
Some comments describing where files are stored will be really helpful.

Reply:
Done.